From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hang Zhou <929513338@qq.com>
Date: Sun, 19 Jul 2026 00:00:00 +0000
Subject: [PATCH] mtd: nand: brcmnand: support ECC strap settings

Allow Broadcom NAND controller users to request the existing strap or
bootloader-programmed ECC settings through the brcm,nand-ecc-use-strap
device-tree property.

Signed-off-by: Hang Zhou <929513338@qq.com>
---
 drivers/mtd/nand/raw/brcmnand/brcmnand.c | 58 +++++++++++++++++++++--
 1 file changed, 55 insertions(+), 3 deletions(-)

--- a/drivers/mtd/nand/raw/brcmnand/brcmnand.c
+++ b/drivers/mtd/nand/raw/brcmnand/brcmnand.c
@@ -960,6 +960,42 @@ static void brcmnand_set_sector_size_1k(
 	nand_writereg(ctrl, acc_control_offs, tmp);
 }
 
+static int brcmnand_get_spare_size(struct brcmnand_host *host)
+{
+	struct brcmnand_controller *ctrl = host->ctrl;
+	u16 acc_control_offs = brcmnand_cs_offset(ctrl, host->cs,
+						  BRCMNAND_CS_ACC_CONTROL);
+	u32 acc = nand_readreg(ctrl, acc_control_offs);
+
+	return acc & brcmnand_spare_area_mask(ctrl);
+}
+
+static void brcmnand_get_ecc_settings(struct brcmnand_host *host,
+				      struct nand_chip *chip)
+{
+	struct brcmnand_controller *ctrl = host->ctrl;
+	u16 acc_control_offs = brcmnand_cs_offset(ctrl, host->cs,
+						  BRCMNAND_CS_ACC_CONTROL);
+	bool sector_size_1k = brcmnand_get_sector_size_1k(host);
+	int spare_area_size = brcmnand_get_spare_size(host);
+	int ecc_level;
+	u32 acc;
+
+	acc = nand_readreg(ctrl, acc_control_offs);
+	ecc_level = (acc & brcmnand_ecc_level_mask(ctrl)) >>
+		    NAND_ACC_CONTROL_ECC_SHIFT;
+
+	if (sector_size_1k)
+		chip->ecc.strength = ecc_level * 2;
+	else if (spare_area_size == 16 && ecc_level == 15)
+		chip->ecc.strength = 1;
+	else
+		chip->ecc.strength = ecc_level;
+
+	if (!chip->ecc.size)
+		chip->ecc.size = sector_size_1k ? 1024 : 512;
+}
+
 /***********************************************************************
  * CS_NAND_SELECT
  ***********************************************************************/
@@ -2280,10 +2316,27 @@ static int brcmnand_setup_dev(struct brc
 	struct brcmnand_cfg *cfg = &host->hwcfg;
 	char msg[128];
 	u32 offs, tmp, oob_sector;
+	bool use_strap;
 	int ret;
 
 	memset(cfg, 0, sizeof(*cfg));
 
+	use_strap = ofnode_read_bool(nand_get_flash_node(chip),
+				     "brcm,nand-ecc-use-strap");
+
+	/*
+	 * Either nand-ecc-xxx or brcm,nand-ecc-use-strap can be set. Error out
+	 * if both exist.
+	 */
+	if (chip->ecc.strength && use_strap) {
+		dev_err(ctrl->dev,
+			"ECC strap and DT ECC configuration properties are mutually exclusive\n");
+		return -EINVAL;
+	}
+
+	if (use_strap)
+		brcmnand_get_ecc_settings(host, chip);
+
 #ifndef __UBOOT__
 	ret = of_property_read_u32(nand_get_flash_node(chip),
 				   "brcm,nand-oob-sector-size",
@@ -2294,9 +2347,8 @@ static int brcmnand_setup_dev(struct brc
 			      &oob_sector);
 #endif /* __UBOOT__ */
 	if (ret) {
-		/* Use detected size */
-		cfg->spare_area_size = mtd->oobsize /
-					(mtd->writesize >> FC_SHIFT);
+		cfg->spare_area_size = use_strap ? brcmnand_get_spare_size(host) :
+			mtd->oobsize / (mtd->writesize >> FC_SHIFT);
 	} else {
 		cfg->spare_area_size = oob_sector;
 	}
